R4R: Bianjie/lcd implementation

HaoyangLiu opened the pull request:
proceed from: https://github.com/cosmos/cosmos-sdk/pull/2118

fedekunze commented on client/lcd/root.go:
```diff
@@ -96,3 +103,100 @@ func createHandler(cdc *wire.Codec) http.Handler {
+// Which is much friendly for further development
+func ServeSwaggerCommand(cdc *wire.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "lite-server",
```
@HaoyangLiu Could you add the cmd to the `Gaia-lite` section of [clients.md](https://github.com/cosmos/cosmos-sdk/blob/develop/docs/sdk/clients.md) docs ?

fedekunze commented on cmd/gaia/cmd/gaiacli/main.go:
```diff
@@ -66,6 +67,7 @@ func main() {
 		tendermintCmd,
 		ibcCmd,
 		lcd.ServeCommand(cdc),
+    lcd.ServeSwaggerCommand(cdc),
```
Please add to Basecoin's [basecli](https://github.com/cosmos/cosmos-sdk/blob/develop/examples/basecoin/cmd/basecli/main.go) file as well

HaoyangLiu commented:
@fedekunze 
I have addressed the issue you mentioned. Besides, I notice that there are some changes about key management APIs in doc/light/api.md. So I change my code to accommodate the changes. In addition, I also refactor the code in multistore proof verification. Previously I rewrite the process to calculate appHash. Just now, I found I can just reuse the commitInfo appHash calculation. So I changed my code to reuse it.

fedekunze commented on docs/light/getting_started.md:
````diff
@@ -1,40 +1,38 @@
+
+Sample command to start gaia-lite node:
+```
+gaiacli lite-server --chain-id {your chain id}
````
Lets change this for `gaiacli lite-server --chain-id=<chain_id>` to keep consistency with the other cmds

fedekunze commented on docs/light/getting_started.md:
````diff
@@ -1,40 +1,38 @@
-### Transfer a token
+When the connected full node is trusted, then the proof is not necessary, so you can run gaia-lite node with trust-node option:
+```
+gaiacli lite-server --chain-id {your chain id} --trust-node
````
same here

fedekunze commented on docs/light/getting_started.md:
````diff
@@ -1,40 +1,38 @@
-![transfer](pics/transfer-tokens.png)
+If you have want to run gaia-lite node in a remote server, then you must specify the public ip to swagger-host-ip
+```
+gaiacli lite-server --chain-id {your chain id} --swagger-host-ip {remove server public ip}
````
`gaiacli lite-server --chain-id=<chain_id> --swagger-host-ip=<remote_server_ip>`. Typo in remove ?

fedekunze commented on docs/sdk/clients.md:
```diff
@@ -378,6 +378,13 @@ With the `pool` command you will get the values for:
-::: tip Note
-🚧 We are actively working on documentation for Gaia-lite.
-:::
+The Gaia-Lite is a light gaiad node. Unlike gaiad full node, it doesn't participate consensus and execute transactions, so it only require very little storage and calculation resource. However, it can provide the same security as a gaia full node.
```
This is a very concise and well written explanation about Gaia-lite, thanks !

fedekunze commented on docs/light/getting_started.md:
````diff
@@ -1,40 +1,38 @@
-operation: [build transaction](api.md#create_transfer---post)
+The gaia-lite node can connect to multiple full nodes. Then gaia-lite will do load balancing for full nodes which is helpful to improve reliability and TPS. You can do this by this command:
+```
+gaiacli lite-server --chain-id {your chain id} --node-list tcp://10.10.10.10:26657,tcp://20.20.20.20:26657
````
`gaiacli lite-server --chain-id=<chain_id> --node-list=tcp://10.10.10.10:26657,tcp://20.20.20.20:26657`. Although not sure if the node list needs to be in between " ". Will double check that

HaoyangLiu replied:
I have double checked that. The node list doesn't need to be in between " "

fedekunze commented on docs/sdk/clients.md:
````diff
@@ -378,6 +378,13 @@ With the `pool` command you will get the values for:
+
+You can start a gaia-lite node with the following command:
+```
+gaiacli lite-server --chain-id {your chain id}
````
gaiacli lite-server --chain-id-<chain id>

HaoyangLiu commented:
I have change the code according to your comment. Please check it.

jackzampolin commented:
@HaoyangLiu Pulling this down today and will get started testing. Can you make sure all the routes and docs here match this issue: https://github.com/cosmos/cosmos-sdk/issues/2113

Looking forward to getting this merged!

fedekunze commented on docs/light/getting_started.md:
```diff
@@ -1,40 +1,38 @@
 
-Then sign the returned transaction byte array with users' private key. Finally broadcast the signed
-transaction. Refer to this link for how to broadcast the signed transaction: [broadcast transaction](api.md#create_transfer---post)
+The gaia-lite support modular rest APIs. Now it supports four modules: general, key, token and stake. If you need all of them, just start it with this command:
```
@jackzampolin : 
```
Gaia-lite is built in a modular format. Each Cosmos module defines it's own RPC API. Currently the following modules are supported: transaction, key, bank (token), staking, governance, and slashing.
```

HaoyangLiu commented:
Hi @jackzampolin, now ICS0, ICS22, ICS23 are still not supported. The following APIs in ICS1 are not supported too.
```
 GET /auth/accounts/{address}
 POST /keys/{name}/sign
```
Do you expect full implementation of all APIs in https://github.com/cosmos/cosmos-sdk/issues/2113 in this PR? If so, I will continus to implement the unsupported APIs and append the code to this PR.

jackzampolin commented:
@HaoyangLiu I expect full implementation of of `ICS0`, `ICS22`, `ICS23`, but maybe not in this PR. How about you finish up the `ICS0` and `ICS1` work in this PR and implement `ICS0`, `ICS22` and `ICS23` in a follow up PR. Does that work for you?

HaoyangLiu commented:
@jackzampolin 
That's OK for me. I will finish ICS0 and ICS1 ASAP. Thanks

HaoyangLiu commented:
@jackzampolin 
According to the spec, all rest APIs' response have a wrapper. For example, the response of /keys (get) is like this:
```
{
    "rest api":"1.0",
    "code":200,
    "error":"",
    "result":
		[
            {
                "name":"monkey",
                "address":"cosmosaccaddr1fedh326uxqlxs8ph9ej7cf854gz7fd5zlym5pd",
                "pub_key":"cosmosaccpub1zcjduc3q8s8ha96ry4xc5xvjp9tr9w9p0e5lk5y0rpjs5epsfxs4wmf72x3shvus0t"
            },
            {
                "name":"test",
                "address":"cosmosaccaddr1thlqhjqw78zvcy0ua4ldj9gnazqzavyw4eske2",
                "pub_key":"cosmosaccpub1zcjduc3qyx6hlf825jcnj39adpkaxjer95q7yvy25yhfj3dmqy2ctev0rxmse9cuak"
            }
        ]
}
```
However, currently,  the response of /keys (get) in gaiacli rest-server is like this:
```
[
	{
		"name":"monkey",
		"address":"cosmosaccaddr1fedh326uxqlxs8ph9ej7cf854gz7fd5zlym5pd",
		"pub_key":"cosmosaccpub1zcjduc3q8s8ha96ry4xc5xvjp9tr9w9p0e5lk5y0rpjs5epsfxs4wmf72x3shvus0t"
	},
	{
		"name":"test",
		"address":"cosmosaccaddr1thlqhjqw78zvcy0ua4ldj9gnazqzavyw4eske2",
		"pub_key":"cosmosaccpub1zcjduc3qyx6hlf825jcnj39adpkaxjer95q7yvy25yhfj3dmqy2ctev0rxmse9cuak"
	}
]
```
I wonder if the wrapper is necessary. Because the wrapper may bring in some additional work for user who want to switch from "gaiacli rest-server" to "gaiacli lite-server"

jackzampolin commented:
@HaoyangLiu Lets leave off the wrapper and update the documentation accordingly. You can add the docs update to this PR.

HaoyangLiu commented:
So far, I have implemented all rest APIs in ICS0, ICS1, ICS20 and ICS21. As for ICS22 and ICS23, I will implement them in later PR.

alexanderbez commented on client/context/loadbalancing.go:
```diff
@@ -0,0 +1,46 @@
+package context
```
Seems like this entire needs to be formatted correctly. Also, it should probably be renamed to `client_manager.go`

alexanderbez commented on client/context/loadbalancing.go:
```diff
@@ -0,0 +1,46 @@
+type ClientManager struct {
+	clients []rpcclient.Client
+	currentIndex int
+	mutex sync.RWMutex
```
I think it'll suffice to just have a write only mutex here (I don't see any logic that obtains a read-lock).

HaoyangLiu replied:
Do you mean use sync.Mutex instead?

alexanderbez commented on client/context/loadbalancing.go:
```diff
@@ -0,0 +1,46 @@
+		mgr.currentIndex = 0
+	}
+	return client
+}
```
Missing new line.

alexanderbez commented on client/context/query.go:
```diff
@@ -13,11 +13,19 @@ import (
 // GetNode returns an RPC client. If the context's client is not defined, an
 // error is returned.
 func (ctx CLIContext) GetNode() (rpcclient.Client, error) {
+	if ctx.ClientManager != nil {
```
I can't see the reason to have a both a client manager and a client. Seems like we should only have a client manager which, in the original case, has only a single client. Thoughts?

alexanderbez commented on client/context/query.go:
```diff
@@ -159,6 +167,22 @@ func (ctx CLIContext) BroadcastTxAsync(tx []byte) (*ctypes.ResultBroadcastTx, er
 
+// BroadcastTxSync broadcasts transaction bytes to a Tendermint node
+// synchronously.
+func (ctx CLIContext) BroadcastTxSync(tx []byte) (*ctypes.ResultBroadcastTx, error) {
```
Why do we need this over just `BroadcastTx`?

alexanderbez commented on client/context/query.go:
```diff
@@ -281,6 +305,50 @@ func (ctx CLIContext) ensureBroadcastTx(txBytes []byte) error {
 }
 
+// proofVerify perform response proof verification
+func (ctx CLIContext) proofVerify(path string, resp abci.ResponseQuery) error {
```
`verifyProof` sounds better I think.

alexanderbez commented on client/context/query.go:
```diff
@@ -281,6 +305,50 @@ func (ctx CLIContext) ensureBroadcastTx(txBytes []byte) error {
+	}
+
+	// Validate the substore commit hash against trusted appHash
+	substoreCommitHash, err :=  store.VerifyMultiStoreCommitInfo(multiStoreProof.StoreName, multiStoreProof.CommitIDList, commit.Header.AppHash)
```
I think we can break this call into multiple lines for legibility.

alexanderbez commented on client/context/query.go:
```diff
@@ -300,10 +368,15 @@ func (ctx CLIContext) query(path string, key cmn.HexBytes) (res []byte, err erro
 	}
 
 	resp := result.Response
-	if !resp.IsOK() {
```
Is `resp.IsOK()` invalid?

HaoyangLiu replied:
Sorry. Maybe it is a mistake in code merging. I have changed it to resp.IsOK which seems more elegant

alexanderbez commented on client/context/query.go:
```diff
@@ -281,6 +305,50 @@ func (ctx CLIContext) ensureBroadcastTx(txBytes []byte) error {
+// proofVerify perform response proof verification
+func (ctx CLIContext) proofVerify(path string, resp abci.ResponseQuery) error {
+	// Data from trusted node or subspace query doesn't need verification
+	if ctx.TrustNode || !isQueryStoreWithProof(path) {
```
I think the `!isQueryStoreWithProof(path)` should be moved outside to the caller. This provides a better abstraction.

alexanderbez commented on client/context/query.go:
```diff
@@ -313,3 +386,21 @@ func (ctx CLIContext) queryStore(key cmn.HexBytes, storeName, endPath string) ([
+	if len(paths) != 3 {
+		return false
+	}
+	// WARNING This should be consistent with query method in iavlstore.go
```
Can you elaborate on this warning and why we need it? What needs to be done about it?

HaoyangLiu replied:
In [iavlstore.go](https://github.com/cosmos/cosmos-sdk/blob/24c7cb03d3ec1dd5a9ec2f0215c358db5af70191/store/iavlstore.go#L212), proof will be included in response only when the query path is /store or /key. I concern that this may be changed in iavlstore.go. If there are some changes about proof building in iavlstore.go, then we must change code here to be consistent with the changes.

HaoyangLiu commented:
Please move to https://github.com/cosmos/cosmos-sdk/pull/2177.
I'm so sorry for bring in troubles to all of you.

cwgoes commented:
Closing in favor of #2177 (as intended @HaoyangLiu?)
